### PR TITLE
Run tests in parallel and native MacOS build support

### PR DESCRIPTION
**Pull request by janhoy**
This PR runs tests in parallel by starting a process per version. We use GNU parallel tool for this.

It is much nicer with native macOS platform support for building the images than resolving to Vagrant, so this PR also adapts the scripts to work with MacOS, and adds documentation on how to prepare your Mac with dependencies.

### CODE REVIEW

**Comment by janhoy** [comment]
Ok, tests are not yet working due to some setfacl trickery, trying to work around it

**Comment by janhoy** [comment]
Ok, this got a bit more involved, as I started to add parallel test support in the same branch... On my mac I test 4 versions in parallel, spinning up 4 dockers at a time. I have given Docker plenty of CPU and RAM... Seems to work well when I just started using unique folder names for temp folders created by the tests.

**Comment by janhoy** [comment]
I have successfully run all tests in parallel on macOS without errors.

**NOTE** I have not yet tested the scripts related to pushing a release on macOS so there may still be some work to do there

**Comment by janhoy** [comment]
I think this is done now. I ran all tests in 5.5 minutes with 10 parallell processes :)

@makuk66 Please review

PS: Note that we still will need #268 for those on Windows... This PR may generate conflicts in the docs with that pr...

**Comment by janhoy** [comment]
@dsmiley I believe you are on a Mac? If you have time, could you try out the description for macOS in this PR, i.e. install the prerequisites via brew and see if you can kick off a build/test or if I forgot something?

File: tests/tests-8/empty-varsolr-vol-ramdomuser-rootgroup/test.sh [hunk 1]
```diff
@@ -23,19 +23,22 @@ echo "Cleaning up left-over containers from previous runs"
 container_cleanup "$container_name"
 container_cleanup "$container_name-copier"
 
-docker volume rm myvarsolr >/dev/null 2>&1 || true
-docker volume create myvarsolr
+myvarsolr="myvarsolr-${container_name}"
+init_myvarsolr 8983 $myvarsolr
```
[the remark this answers is not in the document]
Comment (janhoy) [reply tests/tests-8/empty-varsolr-vol-ramdomuser-rootgroup/test.sh]: Hmm, I can try without

File: update.md [hunk 2]
```diff
@@ -77,25 +77,38 @@ We don't typically announce the availability of new images.
...
 sudo adduser $USER docker
 ```
 
+### Setting up environment on MacOS
+
+Using [Homebrew](https://brew.sh/), install the necessary dependencies for MacOS:
+
+```bash
+brew install coreutils wget gpg gawk shellcheck git bash parallel
+brew cask install docker
```
[the remark this answers is not in the document]
Comment (janhoy) [reply update.md]: No you don't, so better describe in text that you need Docker installed (e.g. through brew cask install docker)

File: update.md [hunk 3]
```diff
@@ -77,25 +77,38 @@ We don't typically announce the availability of new images.
...
+sudo wget -nv --output-document=/usr/local/bin/bashbrew https://doi-janky.infosiftr.net/job/bashbrew/lastSuccessfulBuild/artifact/bin/bashbrew-darwin-amd64
+sudo chmod a+x /usr/local/bin/bashbrew
+# Make gnu readlink the default. You may wish to undo this after building
```
[the remark this answers is not in the document]
Comment (janhoy) [reply update.md]: I know, but alternative is less user friendly, e.g. `ln -s /path/to/greadlink $HOME/bin/readlink` followed by a `./macinit.sh` which puts HOME/bin first in path... With proper warning folks can do it however they like - it's for developers after all, not?

File: tools/test_all.sh [hunk 4]
```diff
@@ -1,14 +1,30 @@
...
+  sudo ls . > /dev/null
+fi
 full_version_variants=$(awk --field-separator ':' '{print $2}' < "$TOP_DIR/TAGS")
-for full_version_variant in $full_version_variants; do
-  ./tests/test.sh "$full_version_variant"
-done
+rm -rf tests/logs/*
+echo Going to execute tests for these versions: $full_version_variants
+echo Executing tests in parallell with $num_processes processes
+parallel --bar --halt-on-error 1 --jobs $num_processes ./tests/test.sh {} <<< "$full_version_variants" ">" tests/logs/test-{}.log "2>&1"
```
[the remark this answers is not in the document]
Comment (janhoy) [reply tools/test_all.sh]: Yea, that's really annoying. I know there are alternatives out there but this was the first I tried.

File: tests/tests-8/gosu/test.sh [hunk 5]
```diff
@@ -3,6 +3,12 @@
 # A simple test of gosu. We create a myvarsolr, and chown it
 #
 
+if [[ "$OSTYPE" == "darwin"* ]]; then
+  # TODO: Fix this test on Mac
```
[the remark this answers is not in the document]
Comment (janhoy) [reply tests/tests-8/gosu/test.sh]: Don't know how right now and should not be a showstopper for committing initial mac support. Opened #275 to not forget it...

File: tests/tests-8/user_volume/test.sh [hunk 6]
```diff
@@ -23,43 +23,49 @@ echo "Cleaning up left-over containers from previous runs"
...
 
+myvarsolr="myvarsolr-${container_name}"
+init_myvarsolr 8983 $myvarsolr
+mylogs="mylogs-${container_name}"
+init_myvarsolr 8983 $mylogs
+myconf="myconf-${container_name}"
+configsets="configsets-${container_name}"
+
 # create a core by hand:
-rm -fr myconf configsets
```
[the remark this answers is not in the document]
Comment (janhoy) [reply tests/tests-8/user_volume/test.sh]: Think I got them all, please check

File: update.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (janhoy) [reply update.md]: Gave it a try

File: tests/test.sh [hunk 7]
```diff
@@ -26,14 +26,13 @@ else
 fi
 
 echo "Running all tests for $tag"
-find "$test_dir" -mindepth 1 -maxdepth 1 -type d | sed -E -e 's/^\.\///' > tests_to_run
-while read  -r d; do
+
+for d in $(find "$test_dir" -mindepth 1 -maxdepth 1 -type d | sed -E -e 's/^\.\///'); do
```
Comment (janhoy) [inline tests/test.sh]: I did this improvement too as part of this commit - avoids leaving tmp files in the file system after a run.

File: tests/shared.sh [hunk 8]
```diff
@@ -77,10 +61,36 @@ function wait_for_server_started {
...
+    sleep 2
   done
-  printf '\nserver started\n'
+  echo "Server started"
   rm "$log"
-  sleep 4
+  sleep $sleep_time
 }
+
+function init_myvarsolr {
```
Comment (makuk66) [inline tests/shared.sh]: as this is now generalised, it should be renamed `prepare_dir_to_mount` or some such.

File: tests/tests-8/user_volume/test.sh [hunk 9]
```diff
@@ -72,6 +79,6 @@ if ! grep -E -q 'One Dell Way Round Rock, Texas 78682' <<<"$data"; then
 fi
 container_cleanup "$container_name"
 
-rm -fr myconf myvarsolr mylogs
+rm -fr $myconf $myvarsolr $mylogs $configsets
```
Comment (makuk66) [inline tests/tests-8/user_volume/test.sh]: spellcheck wants those to be quoted. If you run:

```
shellcheck -e SC1090,SC2001 $(find . -name '*.sh')                                                                     
```

there are various other recommendations for the changes too. It's a bit of a faff, but worth it.

File: tests/shared.sh (hunk 8 above)
Comment (janhoy) [reply tests/shared.sh]: Done

File: tests/tests-8/user_volume/test.sh (hunk 9 above)
Comment (janhoy) [reply tests/tests-8/user_volume/test.sh]: Done

**Comment by dsmiley** [comment]
BTW I successfully followed the instructions up to and including running tools/test_all.sh
Nice work!
I didn't do steps beyond that which seem more for when we're literally going to do a release.

**Pull request merged**